### PR TITLE
aws-java-sdk-core: EC2MetadataClient.java - Adds sending of a User-Ag…

### PR DESCRIPTION
**Pull request by willbengtson**
Adds the sending of a User-Agent other than default User-Agent in Java when making requests to metadata service

Similar to AWS Golang SDK and now AWS Ruby SDK, send a User-Agent when making requests to metadata for credentials.
Golang uses a structure like aws-sdk-go/1.8.19 and now Ruby uses aws-sdk-ruby3/ as of PR #1762.  PR #1445 adds this for python as well.

This will enable protection of AWS credentials from Server Side Request Foregery (SSRF) vectors by use of a metadata proxy and rejecting User-Agents that do not meet a regex.

### CODE REVIEW

**Comment by varunnvs92** [comment]
Can you provide more details on how change this would prevent SSRF?

**Comment by willbengtson** [comment]
@varunnvs92 sure thing!  By providing a User-Agent that is predictable, you can build a metadata proxy to put on your instance that inspects the User-Agent sent in the request.  If the User-Agent does not match or fit the pattern you are expecting then you can block that request.  

Right now, there is not a reasonable way to do so since the default User-Agent is `Java/<jre version>`. 

With SSRF, the attacker is tricking your application to make the request and if an attacker knows you run in AWS, a common tactic is to pull AWS credentials via SSRF to the metadata service.

**Comment by willbengtson** [comment]
I should add that with a successful SSRF, the request would be coming from the library the application uses for HTTP requests, which if no User-Agent is set would be the same as what the SDK uses currently.

File: aws-java-sdk-core/src/main/java/com/amazonaws/internal/EC2CredentialsUtils.java [hunk 1]
```diff
@@ -76,6 +79,11 @@ public static EC2CredentialsUtils getInstance() {
      *             If the requested service is not found.
      */
     public String readResource(URI endpoint) throws IOException {
+        Map<String, String> headers = new HashMap<String, String>();
+        headers.put("User-Agent", USER_AGENT);
+        headers.put("Accept", "*/*");
+        headers.put("Connection", "keep-alive");
+
         return readResource(endpoint, CredentialsEndpointRetryPolicy.NO_RETRY, new HashMap<String, String>());
```
[the remark this answers is not in the document]
Comment (willbengtson) [reply aws-java-sdk-core/src/main/java/com/amazonaws/internal/EC2CredentialsUtils.java]: good catch.  Will update!

**Comment by varunnvs92** [comment]
I have made the change internally. It will be included in today's release.

**Comment by muhqu** [comment]
@willbengtson, @varunnvs92  It looks like this PR was incomplete. When the SDK fetches credentials for from the EC2 metadata service it does so my making two http requests. 
1. `GET /latest/meta-data/iam/security-credentials/`
  To get the name of the IAM Instance Profile. Here `User-Agent: aws-sdk-java/{version}` is added.
2. `GET /latest/meta-data/iam/security-credentials/some-ec2-iam-role-abcdef`
  To retrieve the credentials. Here the **SDK User-Agent is still missing**.

This is because the [EC2CredentialsFetcher](https://github.com/aws/aws-sdk-java/blob/9d9bc2fb303181b60f404af9a84b2300e8c20550/aws-java-sdk-core/src/main/java/com/amazonaws/auth/EC2CredentialsFetcher.java#L121-L124) is directly supplying the headers map...

https://github.com/aws/aws-sdk-java/blob/9d9bc2fb303181b60f404af9a84b2300e8c20550/aws-java-sdk-core/src/main/java/com/amazonaws/auth/EC2CredentialsFetcher.java#L121-L124

... and therefor does not get the defaults defined here…

https://github.com/aws/aws-sdk-java/blob/9d9bc2fb303181b60f404af9a84b2300e8c20550/aws-java-sdk-core/src/main/java/com/amazonaws/internal/EC2CredentialsUtils.java#L81-L88

I think it would be better to add the default headers here… https://github.com/aws/aws-sdk-java/blob/9d9bc2fb303181b60f404af9a84b2300e8c20550/aws-java-sdk-core/src/main/java/com/amazonaws/internal/EC2CredentialsUtils.java#L110-L113 

WDYT?

**Comment by willbengtson** [comment]
@muhqu I think this makes sense.  Good catch.

**Pull request merged**